identity_document/src/document/core_document.rs

```diff
@@ -865,12 +807,67 @@ where
       None => self.data.verification_method.query_mut(query),
     }
   }
+
+  /// Update the DID components of the document's `id`, controllers, methods and services by applying the provided
+  /// fallible maps.
+  ///
+  /// This is an advanced method that can be useful for DID methods that do not know the document's identifier prior
+  /// to publishing, but should preferably be avoided otherwise.
+  ///
+  /// # Errors
+  /// Any error is returned if any of the functions fail or the updates cause scoped method references to embedded
+  /// methods, or methods and services with identical identifiers in the document. In the case where illegal identifiers
+  /// are detected the supplied the `error_cast` function gets called in order to convert [`Error`] to `E`.
+  pub fn try_update_identifiers<F, G, H, L, M, E>(
+    self,
+    id_update: F,
+    controller_update: G,
+    methods_update: H,
+    service_update: L,
+    error_cast: M,
+  ) -> Result<Self, E>
+  where
+    F: FnOnce(CoreDID) -> std::result::Result<CoreDID, E>,
+    G: FnMut(CoreDID) -> std::result::Result<CoreDID, E>,
+    H: FnMut(CoreDID) -> std::result::Result<CoreDID, E>,
+    L: FnMut(CoreDID) -> std::result::Result<CoreDID, E>,
+    M: FnOnce(crate::Error) -> E,
+  {
+    let data = self
+      .data
+      .try_update_identifiers(id_update, controller_update, methods_update, service_update)?;
+    CoreDocument::try_from(data).map_err(error_cast)
+  }
+
+  /// Unchecked version of [Self::try_update_identifiers](Self::try_update_identifiers()).
+  pub fn update_identifiers_unchecked<F, G, H, L>(
+    self,
+    id_update: F,
+    mut controller_update: G,
+    mut methods_update: H,
+    mut service_update: L,
+  ) -> Self
```

Comment:
Same for this method, do we really need it? The API surface of `CoreDocument` is quite large as it is, I'm just wary of adding more convenience methods.

Reply:
We do use it in `StateMetadataDocument` thus I would prefer to keep it. Would you be open for creating a new module in `identity_document` named (something like) `advanced_api` where we place all of these mapping methods, and maybe also things like `resolve_method_mut`, `id_mut_unchecked` etc. and then we can feature gate that module under an `advanced-api` feature?

Reply:
The main point is that users can build it themselves by calling `try_update_identifiers`, hence why I call it a convenience method. It doesn't access any private fields that only an inherent method could. 
Would prefer to remove it to keep the API surface somewhat in check, but will leave the decision to you.

Reply:
> It doesn't access any private fields that only an inherent method could

It actually does (on the last line) as it returns `CoreDocument` directly and does not go through `CoreDocumentData::try_into::<CoreDocument>`. Users don't have access to this optimization.  I think I would prefer to keep this as is, but I see your point.